include/shark/Models/Trees/General.h

```diff
@@ -0,0 +1,126 @@
+//===========================================================================
+/*!
+ *
+ *
+ * \brief       General functions for Tree modeling.
+ *
+ *
+ *
+ * \author      J. Wrigley
+ * \date        2016
+ *
+ *
+ * \par Copyright 1995-2015 Shark Development Team
+ *
+ * <BR><HR>
+ * This file is part of Shark.
+ * <http://image.diku.dk/shark/>
+ *
+ * Shark is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Shark is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Shark.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+//===========================================================================
+
+#ifndef SHARK_MODELS_TREES_GENERAL_H
+#define SHARK_MODELS_TREES_GENERAL_H
+#include <memory>
+template<class T> class TD;
+#define typeof(param) TD<decltype(param)>{}
+namespace shark {
+namespace detail {
+namespace cart {
+// Helper functions
+template<class T> using sink = T;
+struct fail_fast:std::runtime_error{
+	explicit fail_fast(char const* const msg) :std::runtime_error(msg) {}
+};
+#define  CART_STRINGIFY_DETAIL(num) #num
+#define CART_STRINGIFY(num) CART_STRINGIFY_DETAIL(num)
+#define CART_Expects(cond) \
+	if(!cond){ \
+		throw shark::detail::cart::fail_fast{"expectation failed at " \
+        __FILE__ ": " CART_STRINGIFY(__LINE__)}; \
+	}
+
+#define DEBUG(msg) std::cout << msg << std::endl
+template<class Iterator>
+inline std::size_t argmax(Iterator begin, Iterator end) {
+	std::size_t maxIndex = 0;
+	typename Iterator::value_type max = 0.;
+	for(Iterator i = begin; i<end; ++i){
+		if(*i>max){
+			max = *i;
+			maxIndex = i.index();
+		}
+	}
+	return maxIndex;
+}
+template<class Container>
+inline std::size_t argmax(Container c) {
+	return argmax(std::begin(c),std::end(c));
+}
+
+template<class Container,class F>
+inline void fill_fn(Container &v, F &&f){
+	for(std::size_t i=0, s = v.size(); i<s; ++i) v[i] = f(i);
+};
+
+template<class Container, class F>
+inline void generate_i(std::size_t length, Container & v, F&& f){
+	v.clear();
+	for(std::size_t i=0; i<length; ++i) v.push_back(f(i));
+}
+template<class Container, class F>
+inline Container generate_i(std::size_t const length, F&& f){
+	Container v(length);
+	for(std::size_t i=0; i<length; ++i) v[i] = f(i);
+	return v;
+}
+
+template<class T, class F>
+inline T stable_sum(std::size_t i, std::size_t n, F&& f) {
```

Comment:
we should use a normal sum than. it is not so important that we have high precision as we would assume that any decision based on roundoff error is pretty much random from the start. (most machine learning libraries favour float instead of double as default because of this reason)